### PR TITLE
Improve anod specs and plans loading performance

### PR DESCRIPTION
**Pull request by Nikokrock**
Improve loading performance by avoiding source code loading when getting frame information

### CODE REVIEW

**Review by enzbang — APPROVED** [review]

File: src/e3/anod/loader.py [hunk 1]
```diff
@@ -268,38 +273,25 @@ def load(self, repository: AnodSpecRepository) -> Callable[..., Anod]:
...
-    if fixed_spec_repository is not None:
-        return fixed_spec_repository.load(name)
-
-    for k in inspect.stack()[1:]:
-        if "__spec_repository" in k[0].f_globals:
-            spec_repository = k[0].f_globals["__spec_repository"]
+    spec_repository: AnodSpecRepository | None = None
+
+    # Implementation note: context=0 means that the no source context is
+    # computed for each frame. This improve drastically the performance
```
Comment (setton) [inline src/e3/anod/loader.py:290]:
```suggestion
    # computed for each frame. This improves drastically the performance
```

File: src/e3/electrolyt/plan.py [hunk 2]
```diff
@@ -292,7 +292,12 @@ def _add_action(self, name: str, *args: Any, **kwargs: Any) -> None:
         plan_line = "unknown filename:unknown lineno"
         # Retrieve the plan line
         try:
-            caller_frames = inspect.getouterframes(frame=inspect.currentframe())
+            # Implementation note: context=0 means that the no source context is
+            # computed for each frame. This improve drastically the performance
```
Comment (setton) [inline src/e3/electrolyt/plan.py:296]:
```suggestion
            # computed for each frame. This improves drastically the performance
```

File: src/e3/anod/loader.py [hunk 3]
```diff
@@ -268,38 +273,25 @@ def load(self, repository: AnodSpecRepository) -> Callable[..., Anod]:
...
-    :param spec_repository: the spec repository to set
-    """
-    global fixed_spec_repository
-    fixed_spec_repository = spec_repository
-
-
 def spec(name: str) -> Callable[..., Anod]:
     """Load an Anod spec class.
 
+    Obsolete: keep until all from e3.anod.loader import spec are removed from the
```
Comment (enzbang) [inline src/e3/anod/loader.py:279]: I think it is worth bumping the API version and rejecting calls to spec with the new version.

File: src/e3/anod/spec.py [hunk 4]
```diff
@@ -16,12 +16,19 @@
...
 
-SUPPORTED_API = (__version__, "1.5")
-# The driver can support multiple version of the spec API, we currently support
-# only the version 1.4 and 1.5. Default is still 1.4
+SUPPORTED_API = (__version__, "1.5", "1.6")
+# API VERSIONS
+#
+# Version 1.4 (initial version)
+# Version 1.5
+#    NEW: YAML files are also searched in subdirectories whose basename is the
```
Comment (enzbang) [inline src/e3/anod/spec.py:27]: If the specs repo declares version 1.5 and above, the YAML files will only be searched in subdirectories; see e3.anod.spec.load_config_file

Note that you should create a new entry in NEWS.md.

File: src/e3/anod/loader.py [hunk 5]
```diff
@@ -69,7 +69,21 @@ def __init__(
         if not os.path.isdir(spec_dir):
             raise SandBoxError(f"spec directory {spec_dir} does not exist")
         self.spec_dir = spec_dir
-        self.api_version = __version__
+
+        # Read read the API version file
```
Comment (liaiss) [inline src/e3/anod/loader.py:73]: minor, two read in the comment

File: src/e3/anod/spec.py (hunk 4 above)
Comment (Nikokrock) [reply src/e3/anod/spec.py:27]: Thanks I will review that. In loader.py files were still searched in root dir

File: src/e3/anod/loader.py [hunk 6]
```diff
@@ -69,7 +69,21 @@ def __init__(
...
+        version_file = os.path.join(self.spec_dir, "VERSION")
+        if os.path.isfile(version_file):
+            with open(version_file) as f:
+                content = f.read().strip()
+                if ":" not in content:
+                    raise SandBoxError(
+                        f"invalid VERSION file in spec dir {self.spec_dir}"
+                    )
+                self.api_version = content.split(":")[1].strip()
```
Comment (grouigrokon) [inline src/e3/anod/loader.py:82]: We may check that the API version retrieved this way belongs to the `SUPPORTED_API` tuple ?

File: src/e3/anod/loader.py (hunk 5 above)
Comment (Nikokrock) [reply src/e3/anod/loader.py]: Taken into account

File: src/e3/anod/loader.py (hunk 6 above)
Comment (Nikokrock) [reply src/e3/anod/loader.py:83]: thanks !

**Review by enzbang — APPROVED** [review]

**Pull request merged**